Update the mechanism to mark a xact as remote

poojanilangekar opened the pull request.

poojanilangekar committed:
Update the mechanism to mark a xact as remote

ctring commented on pgxn/remotexact/funcs.c:
```diff
@@ -32,22 +32,22 @@ validate_and_apply_xact(PG_FUNCTION_ARGS)
+	 * isRemoteXact flag in MyProc. We don't lock the ProcArray because its
+	 * our own process.
+	 */
+	MyProc->isRemoteXact = true;
```
The proof for deadlock safety relies on the order of transaction upgrade and edge addition to be observed correctly by everyone. Could you add `pg_write_barrier()` after every assignment to `isRemoteXact` to make sure every memory write after that is ordered after such assignment? 
While it is very unlikely that reordering will happen at all, this rules out that possibility completely.

poojanilangekar replied:
Good catch; for some reason, I assumed this would be a TAS.

ctring commented on pgxn/remotexact/remotexact.c:
```diff
@@ -146,8 +135,15 @@ rwset_add_region(int region)
+	 * process as executing a remotexact while checking for deadlocks, thus
+	 * detecting a potential deadlock. 
+	 */
+	MyProc->isRemoteXact = true;
```
`rwset_add_region` is called for every transaction, including single-region transaction, because we don't know if a single-region will be upgraded to multi-region and we need to record all regions involving in a transaction. So you need to check `region_set` to make sure that there are more than 1 region involved here.

poojanilangekar replied:
Done.

poojanilangekar committed:
Two main fixes to remotexact marking:
Add abarrier after each write to isRemoteXact.
Mark a proc as executing a remotexact only when region != current_region.
